feat: add intent based swap support

oscarwroche opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Adds intent based swap support (uses cowswap provider)

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Adds intent based swap support

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/BETR-5

## **Manual testing steps**

- Go to swap page
- Select RWA token (like AAPLON)
- initiate a swap

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces intent-based swaps and a dedicated submission path, along with UI and data handling updates.
> 
> - New intent flow: `handleIntentTransaction` signs CoW order via `KeyringController:signTypedMessage` (EIP-712 V4) and submits using `BridgeStatusController.submitIntent`; comprehensive tests in `app/lib/transaction/intent.test.ts`
> - Type/submit wiring: adds `BridgeQuoteResponse` (with `aggregator`, `walletAddress`, optional `intent`); `BridgeView` now builds this payload and disables submit without `walletAddress`; `useSubmitBridgeTx` routes quotes with `quote.intent` to the intent handler, otherwise falls back to `submitTx`
> - Transactions UI: map bridge `SUBMITTED` and intent status to list item status; ensure status rendering and segments use the mapped value; `TransactionDetails` recognizes `SUBMITTED` color
> - History linking: `useBridgeTxHistoryData` can resolve items by `originalTransactionId` for intent-based flows
> - Unified tx list: refine submitted dedupe and confirmed filtering—use `nonce` when present, otherwise `actionId`; handle undefined/null nonce; adds targeted tests
> - Dependency bumps: updates `@metamask/bridge-controller` and `@metamask/bridge-status-controller` (and related controllers) to latest compatible versions
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit df55af1b3f38e68c3c74595c74c78e5282031c0a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

dawnseeker8 committed:
First backup of intent base swap with Cowswap

dawnseeker8 committed:
Fix the historyItem status mapping not match error.

dawnseeker8 committed:
Update transaction status handling and bump dependencies in yarn.lock

- Added a new status 'SUBMITTED' to the INTENT_STATUS object.
- Updated the logic in TransactionElement to handle the new 'SUBMITTED' status correctly, defaulting to 'FAILED' for unknown statuses.
- Bumped versions of @metamask/bridge-controller and @metamask/bridge-status-controller in yarn.lock, along with their corresponding checksums and resolutions.
- Updated @metamask/transaction-controller version to 60.9.0 in yarn.lock.

oscarwroche committed:
feat: handle intent transaction case

oscarwroche committed:
chore: update lockfile

oscarwroche committed:
sync

oscarwroche committed:
fix: fix signing for cowswap

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
feat: wip

Akaryatrh committed:
Merge remote-tracking branch 'origin/main' into wip/cowswap-integration-with-approval

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
feat: reset bridge url and update lock file with preview packages

Akaryatrh committed:
fix: issue with renderTxElementIcon params

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
fix: type linting issues

micaelae commented on app/lib/transaction/intent.ts:
```diff
@@ -0,0 +1,188 @@
+      throw new Error('Intent order requires sellAmount or buyAmount');
+    }
+
+    const message = {
```
Why is the message normalized in the client? This should be done in the backend and the clients should ideally be able to use whatever is provided, without having to modify fields

micaelae commented on app/lib/transaction/intent.ts:
```diff
@@ -0,0 +1,188 @@
+  );
+
+  const intent = quoteResponse.quote.intent;
+  if (intent && intent.protocol === 'cowswap') {
```
The backend should be the source of truth for supported protocols and the clients should submit the intent order as-is:

- the `submitOrder` endpoint should reject if the intent's protocol is not supported
- additionally, the `getQuoteStream` and `getQuote` endpoints should only return quotes for supported protocols

```suggestion
  const intent = quoteResponse.quote.intent;
  if (intent) {
```

Akaryatrh replied:
Fixed

Akaryatrh committed:
feat: types alignement and fix some comments from review

Akaryatrh committed:
feat: update todo comment for intent

Akaryatrh committed:
Merge remote-tracking branch 'origin/main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
feat: update lockfile after merge from main

Akaryatrh committed:
feat: update controller packages versions

Akaryatrh committed:
feat: revert added code in transactionElement

Akaryatrh committed:
fix: cursor comment about txStatus access

matthewwalsh0 committed:
Support max amount deposit

matthewwalsh0 committed:
Add resolutions

Akaryatrh committed:
fix: wrong condition when filtering tx for intent based tx

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche committed:
chore: deduplicate packages

Akaryatrh committed:
fix: update BridgeView unit test

Akaryatrh committed:
fix: type issue in useSubmitBridgeTx.test.tsx

Akaryatrh committed:
fix: avoid nonce being considered as false when equal to 0

Akaryatrh committed:
fix: update tests for UnifiedTransactionsView to cover nonce check

Akaryatrh committed:
fix: update tests for useSubmitBridgeTx

Akaryatrh committed:
feat: add tests for intent swap related functions

Akaryatrh committed:
Merge remote-tracking branch 'origin/main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
Merge remote-tracking branch 'origin/main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
fix: update lockfile after merge

Akaryatrh committed:
fix: TransactionDetails type issue

dawnseeker8 committed:
feat: add error handling for missing intent order in transaction processing

dawnseeker8 committed:
feat: add error handling for missing or invalid validTo in intent order

dawnseeker8 committed:
fix: ensure transaction status is only mapped when bridgeTxHistoryItem has a valid status

dawnseeker8 commented on app/components/UI/TransactionElement/index.js:
```diff
@@ -449,35 +463,61 @@ class TransactionElement extends PureComponent {
+      transactionStatus = this.mapIntentStatusToTransactionStatus(
+        bridgeTxHistoryItem.status.status,
+      );
+    }
```
Fix in cursor.

dawnseeker8 committed:
fix: update BridgeView to disable submit button when walletAddress is missing

dawnseeker8 commented on app/components/UI/Bridge/Views/BridgeView/index.tsx:
```diff
@@ -127,6 +128,7 @@ const BridgeView = () => {
   const isHardwareAddress = selectedAddress
     ? !!isHardwareAccount(selectedAddress)
     : false;
+  const walletAddress = useSelector(selectSourceWalletAddress);
```
fix by cursor.

dawnseeker8 committed:
fix: update BridgeView test to check button disabled state directly

cmd-ob committed:
fix: Update smart E2E selection fallback when job fails (#24460)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

Fix Smart E2E Selection job failure blocking PR builds and tests by
adding a fallback

   ### Changes:
- Add `final-outputs` step with `if: always()` to guarantee outputs are
set even if AI analysis fails
   - Use `|| echo` to prevent script failures from stopping the action
   - Fix JSON quoting issue using `printf` to preserve `["ALL"]` format
   - Improve user-facing PR comment when AI is unavailable
- Outputs `["ALL"]` with confidence `0` on failure, triggering all
applicable tests

   ### How it works:
   1. AI analysis runs and sets outputs
2. If AI fails, the `final-outputs` step catches it and sets fallback
values
3. Downstream jobs receive guaranteed outputs and run tests accordingly

   ## **Related issues**

- Failing job:
https://github.com/MetaMask/metamask-mobile/actions/runs/20932157126/job/60146624935?pr=22076
## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Strengthens the `smart-e2e-selection` action to avoid blocking builds
when AI analysis fails and to guarantee consumable outputs.
> 
> - New `final-outputs` step (with `if: always()`) sets
`ai_e2e_test_tags` (fallback to `["ALL"]`) and passes through
`ai_confidence` when analysis fails
> - Wraps AI script with `|| echo` so failures don’t break the job;
initializes default outputs in `ai-analysis`
> - Updates outputs to read from `final-outputs` and always displays
them
> - PR comment logic: uses `pr_comment.md` if present; otherwise posts a
clear fallback message; preserves skip messaging
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
c33e87bff9b9eda376b3aa9213630ac290fb2e21. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

matthewwalsh0 committed:
Fix resolutions

pedronfigueiredo committed:
feat: Add simulation receiving value for Predict withdraw metrics (#24281)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Predict withdraw transactions were missing
`simulation_receiving_assets_total_value`, limiting visibility into
total value withdrawn.

Attach `assetsFiatValues.receiving` to Predict withdraw transactions
during signing and add tests to verify the value is emitted through the
metrics pipeline.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/6392

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds missing simulation receiving value for Predict withdraws and
validates metrics propagation.
> 
> - Updates `PredictController.beforeSign` to set updated `txParams` and
`assetsFiatValues.receiving` (stringified amount), and conditionally
apply gas estimates
> - Extends tests: `PredictController.test` asserts
`assetsFiatValues.receiving`; `metrics.test` verifies
`simulation_receiving_assets_total_value` is included when present and
adds coverage for Smart Transactions metrics properties
> 
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
574d483c2fc2877b7deee7a7aacf035152b20486. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

Akaryatrh committed:
fix: deduplicate by actionId if nonce is missing

Akaryatrh committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

matthewwalsh0 committed:
Fix resolutions

matthewwalsh0 committed:
Remove resolution

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche committed:
refactor: refactor tests

oscarwroche committed:
refactor: refactor tests

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche committed:
fix: put back old test

micaelae commented on app/components/UI/TransactionElement/index.js:
```diff
@@ -449,35 +463,61 @@ class TransactionElement extends PureComponent {
     const { value, fiatValue = false, actionKey } = transactionElement;
+    // transform the status from bridgeTxHistoryItem to the status of the transaction
+    let transactionStatus = status;
+    if (bridgeTxHistoryItem && bridgeTxHistoryItem.status) {
```
This needs one more condition so the status is only translated if the txHistory item is an intent

```suggestion
    if (bridgeTxHistoryItem && bridgeTxHistoryItem.status && bridgeTxHistoryItem.quote.intent) {
```

Akaryatrh committed:
fix: filter out non intent bridge tx when mapping intent status

Akaryatrh committed:
fix: remove useless selector for wallet address

oscarwroche committed:
chore: disable any rule for test file

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche committed:
fix: broaden type in test file to fix type error

micaelae commented on app/components/UI/Bridge/types.ts:
```diff
@@ -19,6 +24,13 @@ export interface BridgeToken {
+    intent?: Intent;
+  };
+
 export enum BridgeViewMode {
```
Please remove this redundant type because aggregator, walletAddress and intent are not in use

```suggestion

export enum BridgeViewMode {
```

oscarwroche replied:
It seems like `aggregator` and `walletAddress` are used in the BridgeView (L349-L-353)

```
const quoteResponse: BridgeQuoteResponse = {
  ...activeQuote,
  aggregator: activeQuote.quote.bridgeId,
  walletAddress,
};
```

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche commented on app/lib/transaction/intent.ts:
```diff
@@ -0,0 +1,176 @@
+
+    const chainId = quoteResponse.quote.srcChainId;
+    const verifyingContract =
+      intent.settlementContract ?? '0x9008D19f58AAbd9eD0D60971565AA8510560ab41';
```
@micaelae agreed - this can be part of V2 as discussed last week?

oscarwroche committed:
refactor: remove intent type casts

oscarwroche committed:
Merge branch 'feat/metamask-pay-max' into wip/cowswap-integration-with-approval

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval

oscarwroche committed:
chore: remove core bridge package resolutions

Akaryatrh commented:
@SocketSecurity ignore npm/@metamask/gas-fee-controller

oscarwroche committed:
chore: deduplicate dependencies

Akaryatrh commented:
@SocketSecurity ignore npm/@metamask/multichain-network-controller

oscarwroche committed:
Merge branch 'main' into wip/cowswap-integration-with-approval
